src/commcare_cloud/ansible/plugins/lookup/cached_aws_secret.py

```diff
@@ -0,0 +1,32 @@
+from ansible.plugins.lookup import aws_secret
+
+__metaclass__ = type
+
+DOCUMENTATION = aws_secret.DOCUMENTATION
+RETURN = aws_secret.RETURN
+
+
+class LookupModule(aws_secret.LookupModule):
+    _secrets_cache = {}
```

Comment:
You're adding caching here while the implementation of `AwsSecretsBackend.get_secret` constructs a new session and client on every call.

https://github.com/dimagi/commcare-cloud/blob/f5c96e856c5b0485f9309cb0a173f24781e30310/src/commcare_cloud/environment/secrets/backends/aws_secrets/main.py#L43-L46

Is it possible that that overhead is why it was slow enough that you needed to add caching? For example, is it possible that the client already does caching and this new cache you've added is not needed?

Reply:
I think no, but let me try to explain.

This `cached_aws_secret` plugin is called from ansible, while `AwsSecretsBackend.get_secret` is called from commcare-cloud. In commcare-cloud, we have full control of the code paths and it'll be easy to add caching later if it's needed, but typically `AwsSecretsBackend.get_secret` is called zero to one times within commcare-cloud, and maybe up to a handful in steady state normal operations; ansible on the other hand makes tons of repeat calls to the plugin: once per secret per task per machine. So if there is one task (e.g. copy localsettings) that needs 20 secrets on 12 machines, followed by another task (e.g copy applications.properties for formplayer) that needs a subset of those same secrets, say 10 secrets on 40 machines, then it'll make a total of 20 x 12 + 10 x 40 = 640 calls for secrets. With the caching I do here, we can instead make just 20 calls.